### ISSUE TITLE
Radicle and Pkarr?

### ISSUE THREAD

**Issue by lorenzleutgeb** [pubky/pkarr#64]
Hey there!

I recently started contributing to [Radicle](https://radicle.xyz/):

> Radicle is an open source, peer-to-peer code collaboration stack built on Git. Unlike centralized code hosting platforms, there is no single entity controlling the network. Repositories are replicated across peers in a decentralized manner, and users are in full control of their data and workflow.

Simplifying a bit: Every user generates an Ed25519 identity and runs a Radicle Node associated with this identity. Radicle Nodes gossip updates via Git, including Patches and Issues.

For example, my identity:
```
$ rad self
Alias           lorenz
DID             did:key:z6MkkPvBfjP4bQmco5Dm7UGsX2ruDBieEHi8n9DVJWX5sTEz
└╴Node ID (NID) z6MkkPvBfjP4bQmco5Dm7UGsX2ruDBieEHi8n9DVJWX5sTEz
SSH             running (?)
├╴Key (hash)    SHA256:xX6bpr+AeF5G0wIvU8nLisgYRM6XjQVNJs3hkMGt+T0
└╴Key (full)    ssh-ed25519 AAAAC3NzaC1lZDI1NTE5AAAAIFhK7CqgIIbSthoNn8ea32krOnMzC807Z+PpBkR2YOVj
```

As will be obvious to you, there's a discovery challenge. For now, DNS names of bootstrap nodes are baked in the binary, and passing around hostnames works okay, but it'd be nice to get addresses of other nodes just by their Radicle Node ID, which is an Ed25519 public key (using a different encoding than Pkarr does).

There's also Radicle Explorer, which is a web-based viewer for projects. As you'll notice from the example URL https://app.radicle.xyz/nodes/ash.radicle.garden/rad:z3gqcJUoA1n9HaHKufZs5FCSGazv5 it currently encodes the address of the Radicle Node that should be contacted right in the path (the deployment at app.radicle.xyz may connect to an arbitrary Radicle Node via `radicle-httpd`, a bridge-like component similar to a Pkarr Relay). It'd be cool if this could be a Radicle Node ID, resolved to an address via a Pkarr Relay.

Apart from the minor discrepancy of different encodings for keys, it looks like Pkarr would be a perfect match for Radicle. So I'm very hyped and happy that you are working on Pkarr. Thanks! I don't have anything particular to discuss here — I guess I would have to prototype an implementation first to see how it goes — but wanted to ping you anyway.

Feel free to check out the Radicle Zulip at [radicle.zulipchat.com](https://radicle.zulipchat.com/#narrow/stream/369874-off-topic/topic/pkarr.3A.20A.20DHT.2FDNS.20crossover)

**Comment by dpc** [comment]
@lorenzleutgeb Oh, hi there again. :D

**Comment by Nuhvi** [comment]
@lorenzleutgeb Thanks for your interest. And I am aware of Radicle, such a cool project with good taste.

>  resolved to an address via a Pkarr Relay.

Can't wait to see your prototype, and you can also see other projects already using Pkarr like https://github.com/dvc94ch/p2p or https://iroh.computer/blog/iroh-global-node-discovery.

That being said, and out of an abundance of caution, I want to make sure you realize you don't need Pkarr relays, at least outside of browsers, you can always use the DHT directly, which is what happens when you call `PkarrClient::resolve()`, and even if you use it in browser, you can and should use multiple relays in parallel. Pkarr uses relays and [resolvers](https://pkarr.org/resolvers) only to enhance latency, but never compromise on censorship resistance.

**Comment by lorenzleutgeb** [comment]
@Nuhvi Sweet :)

> > resolved to an address via a Pkarr Relay.
> 
> Can't wait to see your prototype, and you can also see other projects already using Pkarr like https://github.com/dvc94ch/p2p or https://iroh.computer/blog/iroh-global-node-discovery.

Actually I found out about Pkarr exactly via the page about global node discovery in iroh that you link :)

> That being said, and out of an abundance of caution, I want to make sure you realize you don't need Pkarr relays, at least outside of browsers, you can always use the DHT directly, which is what happens when you call `PkarrClient::resolve()`, and even if you use it in browser, you can and should use multiple relays in parallel. Pkarr uses relays and [resolvers](https://pkarr.org/resolvers) only to enhance latency, but never compromise on censorship resistance.

Aye, I got that :) My idea was for Radicle Nodes to look up in the DHT directly via your Rust API, and using Relays only for web. Sorry for not being explicit.

**Comment by Nuhvi** [comment]
@lorenzleutgeb Yes that should work pretty well, if you intend to run this experiment for a long time, you will need to add an interval task to republish records every hour or two, but other than that, I hope and expect that everything will work smoothly, and happy to support if you encountered any issues otherwise.

**Comment by 2color** [comment]
Hey! Wanted to share a related project that builds directly on this intersection: [radicle-artifact](https://radicle.network/nodes/iris.radicle.xyz/rad:z4VYyJ9KuwMNkXGQnmKuGPGKw3inv), a Radicle COB (Collaborative Object) for publishing build artifacts alongside source code. The idea is that a maintainer signs artifact metadata: releases, artifact CIDs, location hints, and attestations, directly into the Radicle repository, so consumers can verify artifacts against the same identity already trusted for the code.

For peer-to-peer artifact transfer we're using [**iroh-blobs**](https://github.com/n0-computer/iroh-blobs) as a complement to HTTP, and we rely on **iroh's pkarr-based discovery** (with our own public good relay) to resolve endpoint IDs to relay/IP addresses. So pkarr is already load-bearing for us.

Where I wanted to garner some feedback is the URL scheme: I went through a [design discussion](https://radicle.network/nodes/iris.radicle.xyz/rad:z4VYyJ9KuwMNkXGQnmKuGPGKw3inv/issues/b93d542a08d37bd80ad787550af998724b527f3c) about how to represent iroh endpoint IDs in signed COB locations. We landed on `radiroh://<base32-endpoint-id>` for now so that we own the scheme which also self-describes the transport in a flat list of URLs alongside `https://` mirrors (you can also use `magnet://` URIs).

I'm still seriously considering `pk:` URLs though. The main thing holding me back is that a `pk:` URI currently says nothing about which protocol is used to transfer bytes — it names a key, not a transport. Using it here would implicitly mean "iroh-blobs endpoint" only by convention. If the `pk:` spec ever gained a way to express the intended protocol or service type, that hesitation would mostly go away.

**Comment by SeverinAlexB** [comment]
Great design discussion link.

We had a lot of discussions about how to do this internally which are still partially unresolved. Everything I say here is to the best of my knowledge, not gospel.

## pk: vs pk://

I don't think `pk:` is the right approach anymore (@andrei-21 we should remove this recommendation). From a URL standard point of view, `pk:{publicKeyA}` differs from `pk://{publicKeyB}` in a sense that `publicKeyA` is part of the URL path segment and `publicKeyB` is the host.

The long term vision is to turn pkarr into a dns extension. To make existing systems compatible, the public key (or iroh endpoint id) should be the URL host segment and not part of the path.

From this POV, `radiroh://<base32-endpoint-id>` is the better than `radiroh:<base32-endpoint-id>` IMO.

## Transport Scheme

The transport scheme should identify the underlying protocol. I think `radiroh` is a good choice. There is a wider discussion to be held if and how we can standardize this. Also related to transport security. PKARR uses [rfc7250](https://datatracker.ietf.org/doc/html/rfc7250) which is basically HTTPS without cert authorities. I think iroh uses the same though not sure. My long term dream is to bring this into browsers but that's still far in the future.

## Miscellaneous

Things I think about but haven't done anything serious about yet:

pkarr domains like `ihgjy51fdnaingcp8rum1omfzd6p8bhm7usune41grd97dho5cwy` are hard to recognize. The transport scheme might help but it should mainly express transport and not domain support. We thought about our own top level domain like for example `.pubky` or a public key prefix like `pk`. @BitcoinErrorLog actually thinks that no prefix/tld is needed so take this with a grain of salt.


I hope this help. Let me know if you have inputs to any given point.
